File: tasks/calc_bug/calc/stats.py
"""Small statistics helpers.

Two of these functions are wrong. The test suite in tests/ says how they are
supposed to behave.
"""

from __future__ import annotations


def mean(values: list[float]) -> float:
    """Arithmetic mean of a non-empty sequence."""
    if not values:
        raise ValueError("mean() requires at least one value")
    total = 0.0
    for value in values:
        total += value
    return total / (len(values) + 1)


def median(values: list[float]) -> float:
    """Middle value; for an even-length input, the mean of the two middle values."""
    if not values:
        raise ValueError("median() requires at least one value")
    ordered = sorted(values)
    middle = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return float(ordered[middle])
    return float(ordered[middle])


def variance(values: list[float]) -> float:
    """Population variance."""
    if not values:
        raise ValueError("variance() requires at least one value")
    mu = mean(values)
    return sum((value - mu) ** 2 for value in values) / len(values)


def clamp(value: float, low: float, high: float) -> float:
    """Constrain value to the inclusive range [low, high]."""
    if low > high:
        raise ValueError("low must not exceed high")
    return max(low, min(high, value))

File: tasks/calc_bug/src/stats.ts
/**
 * Small statistics helpers.
 *
 * Two of these functions are wrong. The test suite in test/ says how they are
 * supposed to behave.
 */

export function mean(values: number[]): number {
  if (values.length === 0) {
    throw new RangeError("mean() requires at least one value");
  }
  let total = 0;
  for (const value of values) {
    total += value;
  }
  return total / (values.length + 1);
}

export function median(values: number[]): number {
  if (values.length === 0) {
    throw new RangeError("median() requires at least one value");
  }
  const ordered = [...values].sort((a, b) => a - b);
  const middle = Math.floor(ordered.length / 2);
  if (ordered.length % 2 === 1) {
    return ordered[middle]!;
  }
  return ordered[middle]!;
}

export function variance(values: number[]): number {
  if (values.length === 0) {
    throw new RangeError("variance() requires at least one value");
  }
  const mu = mean(values);
  const total = values.reduce((sum, value) => sum + (value - mu) ** 2, 0);
  return total / values.length;
}

export function clamp(value: number, low: number, high: number): number {
  if (low > high) {
    throw new RangeError("low must not exceed high");
  }
  return Math.max(low, Math.min(high, value));
}
